vochain/indexer/indexer.go

```diff
@@ -249,37 +254,112 @@ func (idx *Indexer) RestoreBackup(path string) error {
 	return nil
 }
 
-// SaveBackup backs up the database to a file on disk.
+// ImportBackup restores the database from a backup created via ExportBackup.
+// Note that this must be called with ExpectBackupRestore set to true,
+// and before any indexing or queries happen.
+func (idx *Indexer) ImportBackup(r io.Reader) error {
+	if idx.readWriteDB != nil {
+		panic("Indexer.RestoreBackup called after the database was initialized")
+	}
+	log.Debugf("restoring indexer backup")
+	gzipReader, err := gzip.NewReader(r)
+	if err != nil {
+		return fmt.Errorf("could not create gzip reader: %w", err)
+	}
+	defer gzipReader.Close()
+	if err := restoreDBFromSQLDump(idx.dbPath, gzipReader); err != nil {
+		return fmt.Errorf("could not restore indexer backup: %w", err)
+	}
+	if err := idx.startDB(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func restoreDBFromSQLDump(dbPath string, r io.Reader) error {
+	db, err := sql.Open("sqlite3", fmt.Sprintf("file:%s?mode=rwc&_journal_mode=wal&_txlock=immediate&_synchronous=normal&_foreign_keys=true", dbPath))
+	if err != nil {
+		return fmt.Errorf("could not open indexer db: %w", err)
+	}
+	defer db.Close()
+
+	scanner := bufio.NewScanner(r)
+	var statement strings.Builder
+	for scanner.Scan() {
+		line := scanner.Text()
+		statement.WriteString(line)
+		statement.WriteString("\n")
+
+		if strings.HasSuffix(line, ";") {
+			_, err := db.Exec(statement.String())
+			if err != nil {
+				return fmt.Errorf("failed to execute statement: %s (error: %w)", statement.String(), err)
+			}
+			statement.Reset()
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("error during restore: %w", err)
+	}
+
+	return nil
+}
+
+// ExportBackup backs up the database to a file on disk.
 // Note that writes to the database may be blocked until the backup finishes,
 // and an error may occur if a file at path already exists.
 //
 // For sqlite, this is done via "VACUUM INTO", so the resulting file is also a database.
-func (idx *Indexer) SaveBackup(ctx context.Context, path string) error {
-	_, err := idx.readOnlyDB.ExecContext(ctx, `VACUUM INTO ?`, path)
-	return err
+func (idx *Indexer) ExportBackup(w io.Writer) error {
+	log.Debugf("exporting indexer backup")
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	tmpDB, err := os.CreateTemp("", "indexer*.sqlite3")
+	if err != nil {
+		return fmt.Errorf("could not create tmpdb file: %w", err)
+	}
+	defer func() {
+		if err := os.Remove(tmpDB.Name()); err != nil {
+			log.Warnw("error removing tmpdb file", "path", tmpDB.Name(), "err", err)
```

Comment:
nit: don't log from a func that returns an error. you can return this error via a named error return.

Reply:
but it's a defer,
i'll do `defer os.Remove(tmpDB.Name())`

go-golangci-lint complains about
"Error return value of `os.Remove` is not checked (errcheck)"
but we have 1000s of such unchecked errors in defers anyway :sweat_smile: